### PR TITLE
Stop the AI code review from approving PRs it never actually reviewed

### LINKED ISSUE #174: Code-review fan-out fails: all review sub-agents return no findings object
## Context

The AI code-review Action (`code-review-action`) reviews PRs by fanning out 12 specialized `pr:review:*` sub-agents in parallel, each required to emit a structured findings object that is aggregated into the posted review. On a recent run (PR #172, run 26656745335), all 12 sub-agents failed and the Action posted a hollow `APPROVED` with zero findings — the diff was never actually reviewed. Because this is systemic (every agent, identical error), every PR is affected until fixed.

Related: #142 (closed), #161 (closed), #155 (closed)

## What

Each of the 12 sub-agents finishes with the same error: `Sub-agent structured output did not match the findings schema: (root): Invalid input: expected object, received undefined`. The fan-out then reports `failed_count: 12`, `finding_count: 0`, writes an empty `precomputed-reviews.json` (`{"findings": []}`), and submits an empty approving review. Restore the structured-output contract so sub-agents reliably emit a schema-valid findings object (or, when none, a valid empty `{ "findings": [] }`), and make a fully-failed fan-out NOT post a clean approval.

## Why

The review is the repository's primary automated quality gate, and this repo auto-merges approved + green PRs. An empty `APPROVED` from a 100%-failed fan-out gives false confidence and lets unreviewed code merge automatically — strictly worse than a visible failure.

## Scope

- **In scope:**
  - Diagnose why sub-agents return `undefined` at the schema root instead of a findings object (text-only output vs StructuredOutput tool call; schema/contract drift on `@main`)
  - Fix the sub-agent ↔ findings-schema contract in `reviewFanout.ts` / `reviewFindings.ts` / `reviewOutput/reviewOutput.ts` / `runClaude.ts`
  - When all agents fail, surface it as a failed/neutral review instead of an empty approval
  - Add a regression test asserting a valid findings object (incl. empty) and that an all-failed fan-out does not approve
- **Out of scope:**
  - Metrics/run-summary accuracy (filed separately)
  - Reworking the set of review dimensions or models

## Solution

Reproduce against PR #172's head, capture one sub-agent's raw output, and determine whether the model returns text instead of calling the StructuredOutput tool or whether the expected schema shape changed. Re-align the sub-agent prompt/tool contract so the root is always the findings object, normalize a missing object to `{ "findings": [] }` only when the agent genuinely succeeded, and change aggregation so `failed_count === agent_count` yields a non-approving outcome. Lock it in with tests in `reviewFanout.test.ts`.

### PR DESCRIPTION
**Pull request by awinogradov**
The AI code-review Action fans out 12 sub-agents whose structured findings never reached the orchestrator, so it posted an empty APPROVED without reviewing the diff — and this repo auto-merges approved PRs. This recovers the findings and makes a fully-failed review block instead of approve.

- Recover each sub-agent's findings from its final text output when the SDK structured-output channel is empty, still preferring the structured channel when present
- Fail the review check when every sub-agent errors, instead of writing an empty findings file the root model reads as an approval
- Add regression tests for text recovery, empty findings, structured-output precedence, and the all-failed safety net

---

**Release notes:**

- Fix the AI code review posting an empty approval when its review sub-agents return no findings
- A fully-failed review now fails the check instead of silently approving

---

**Issues:**

Closes #174

### CODE REVIEW

File: .github/actions/code-review-action/src/reviewFanout.test.ts [hunk 1]
```diff
@@ -214,6 +215,91 @@ describe("collectStructuredFindings", () => {
     expect(result.error).toContain("findings");
     expect(result.raw).toBe(JSON.stringify({ findings: "nope" }));
   });
+
+  test("recovers findings from the result text when structured_output is absent", async () => {
```
[the remark this answers is not in the document]
Comment (awinogradov) [reply .github/actions/code-review-action/src/reviewFanout.test.ts:219]: @symbiot-bot Fixed — added a test that passes structured_output: null, exercising the null fallback to result text.

File: .github/actions/code-review-action/src/reviewFanout.test.ts [hunk 2]
```diff
@@ -214,6 +215,91 @@ describe("collectStructuredFindings", () => {
...
+  test("recovers a valid empty findings object from the result text", async () => {
+    const result = await collectStructuredFindings(
+      noopLog,
+      streamOf(resultMessage({ subtype: "success", result: '{ "findings": [] }' })),
+    );
+    expect(result).toEqual({ findings: [] });
+    expect(result.error).toBeUndefined();
+  });
+
+  test("strips a fenced code block around the result text", async () => {
```
[the remark this answers is not in the document]
Comment (awinogradov) [reply .github/actions/code-review-action/src/reviewFanout.test.ts:238]: @symbiot-bot Fixed — added a test with prose-wrapped JSON (Result: { ... } done) that exercises the brace-slice recovery strategy.

File: .github/actions/code-review-action/src/reviewFanout.ts [hunk 3]
```diff
@@ -276,12 +276,37 @@ async function runSubagent(
 }
 
 /**
- * Drain the SDK stream and parse the terminal result's structured output.
+ * Recover the findings object an agent emits as its final text when the SDK
```
[the remark this answers is not in the document]
Comment (awinogradov) [reply .github/actions/code-review-action/src/reviewFanout.ts:279]: @symbiot-bot Considered — kept it local. parseJsonWithSchema in reviewOutput.ts does schema-validated single-parse for verdict/reaction output; parseResultJson does tolerant multi-strategy recovery (fence, brace) returning unknown for fan-out text. They overlap only on a trivial JSON.parse, and this helper has one caller, so moving it would add coupling without real reuse.

File: .github/actions/code-review-action/src/submitReview.ts [hunk 4]
```diff
@@ -76,6 +76,15 @@ async function cleanupBotThreads(
 }
 
 // Main execution
+
+// A wholly-failed review fan-out has no usable signal. Fail the check loudly
+// instead of posting a review — guarded before any parse/dedup so a prior
+// same-SHA review cannot flip this green.
+if (process.env.FANOUT_WHOLLY_FAILED === "true") {
```
[the remark this answers is not in the document]
Comment (awinogradov) [reply .github/actions/code-review-action/src/submitReview.ts:83]: @symbiot-bot Considered — submitReview.ts is a top-level script that uses console.log/console.error throughout (including the existing no-structured-output fatal path); it has no pino logger. console.error here matches the file convention, so routing one path through pino would be inconsistent.

File: .github/actions/code-review-action/src/reviewFanout.ts [hunk 5]
```diff
@@ -276,12 +276,37 @@ async function runSubagent(
...
+function parseResultJson(result: string | undefined): unknown {
+  if (!result) return undefined;
+  const trimmed = result.trim();
+  const unfenced = trimmed.replace(/^```(?:json)?\s*\n?/i, "").replace(/\n?```$/i, "").trim();
+  const sliced = trimmed.slice(trimmed.indexOf("{"), trimmed.lastIndexOf("}") + 1);
+  for (const candidate of [trimmed, unfenced, sliced]) {
+    if (!candidate) continue;
+    try {
+      return JSON.parse(candidate);
+    } catch {
```
[the remark this answers is not in the document]
Comment (awinogradov) [reply .github/actions/code-review-action/src/reviewFanout.ts:293]: @symbiot-bot Considered — the empty catch is intentional fall-through to the next recovery strategy (a miss is expected until one parses) and is already commented. parseResultJson is a pure helper with no logger, so per-candidate debug logging would mean threading a logger in for low signal.
